sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -453,6 +453,14 @@ case class Or(left: Expression, right: Expression) extends BinaryOperator with P
 
 abstract class BinaryComparison extends BinaryOperator with Predicate {
 
+  override def inputType: AbstractDataType = AnyDataType
```

Comment:
We have to write the code comments for explaining it. Given this assumption, `inputType` might be misused in the future. 

In addition, this PR has a pretty critical change. We really need to check all the new data types we support in this PR.  https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala#L90-L97

Could you write a comprehensive test case coverage for this?